pkg/apis/gcp/cache/v1alpha1/cloudmemorystore_instance_types.go

```diff
@@ -0,0 +1,238 @@
+
+	// CurrentLocationID is the current zone where the Redis endpoint is placed.
+	// For Basic Tier instances, this will always be the same as the locationId
+	// provided by the user at creation time. For Standard Tier instances, this
+	// can be either locationId or alternativeLocationId and can change after a
+	// failover event.
+	CurrentLocationID string `json:"currentLocationId,omitempty"`
+
+	// Endpoint of the Cloud Memorystore instance used in connection strings.
+	Endpoint string `json:"endpoint,omitempty"`
+
+	// Port at which the Cloud Memorystore instance endpoint is listening.
+	Port int `json:"port,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CloudMemorystoreInstance is the Schema for the instances API
+// +k8s:openapi-gen=true
+// +groupName=cache.gcp
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.state"
+// +kubebuilder:printcolumn:name="CLASS",type="string",JSONPath=".spec.classRef.name"
+// +kubebuilder:printcolumn:name="VERSION",type="string",JSONPath=".spec.redisVersion"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+type CloudMemorystoreInstance struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CloudMemorystoreInstanceSpec   `json:"spec,omitempty"`
+	Status CloudMemorystoreInstanceStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CloudMemorystoreInstanceList contains a list of CloudMemorystoreInstance
+type CloudMemorystoreInstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []CloudMemorystoreInstance `json:"items"`
+}
+
+// NewCloudMemorystoreInstanceSpec creates a new CloudMemorystoreInstanceSpec
+// from the given properties map.
+func NewCloudMemorystoreInstanceSpec(properties map[string]string) *CloudMemorystoreInstanceSpec {
+	spec := &CloudMemorystoreInstanceSpec{
+		ReclaimPolicy: corev1alpha1.ReclaimRetain,
+
+		// Note that these keys should match the JSON tags of their respective
+		// CloudMemorystoreInstanceSpec fields.
+		Region:                properties["region"],
+		Tier:                  properties["tier"],
+		LocationID:            properties["locationId"],
+		AlternativeLocationID: properties["alternativeLocationId"],
+		ReservedIPRange:       properties["reservedIpRange"],
+		AuthorizedNetwork:     properties["authorizedNetwork"],
+		RedisVersion:          properties["redisVersion"],
+		RedisConfigs:          parseMap(properties["redisConfigs"]),
+	}
+
+	if i, err := strconv.ParseInt(properties["memorySizeGb"], 10, 0); err == nil {
+		spec.MemorySizeGB = i
+	}
+
+	return spec
+}
+
+// parseMap parses a string of comma separated key value pairs, for example
+// "key1: value1, key2: value2", into a string map.
+func parseMap(s string) map[string]string {
```

Comment:
this parseMap function could just be defined in a single place in the util package

Reply:
See my comment on `parseSlice`.